src/web-experiments.ts

```diff
@@ -0,0 +1,261 @@
+        return window?.location
+    }
+
+    private static matchUTMConditions(testVariant: WebExperimentVariant): boolean {
+        if (isNullish(testVariant.conditions) || isNullish(testVariant.conditions?.utm)) {
+            return true
+        }
+        const campaignParams = Info.campaignParams()
+        if (campaignParams['utm_source']) {
+            // eslint-disable-next-line compat/compat
+            const utmCampaignMatched = testVariant.conditions?.utm?.utm_campaign
+                ? testVariant.conditions?.utm?.utm_campaign == campaignParams['utm_campaign']
+                : true
+
+            const utmSourceMatched = testVariant.conditions?.utm?.utm_source
+                ? testVariant.conditions?.utm?.utm_source == campaignParams['utm_source']
+                : true
+
+            const utmMediumMatched = testVariant.conditions?.utm?.utm_medium
+                ? testVariant.conditions?.utm?.utm_medium == campaignParams['utm_medium']
+                : true
+
+            const utmTermMatched = testVariant.conditions?.utm?.utm_term
+                ? testVariant.conditions?.utm?.utm_term == campaignParams['utm_term']
+                : true
+
+            return utmCampaignMatched && utmMediumMatched && utmTermMatched && utmSourceMatched
+        }
+
+        return false
+    }
+
+    private static logInfo(msg: string, ...args: any[]) {
+        logger.info(`[WebExperiments] ${msg}`, args)
+    }
+
+    private static applyTransforms(experiment: string, variant: string, transforms: WebExperimentTransform[]) {
+        transforms.forEach((transform) => {
+            if (transform.selector) {
+                WebExperiments.logInfo(
+                    `applying transform of variant ${variant} for experiment ${experiment} `,
+                    transform
+                )
+                // eslint-disable-next-line no-restricted-globals
+                const elements = document?.querySelectorAll(transform.selector)
+                elements?.forEach((element) => {
+                    const htmlElement = element as HTMLElement
+                    if (transform.attributes) {
+                        transform.attributes.forEach((attribute) => {
+                            switch (attribute.name) {
+                                case 'text':
+                                    htmlElement.innerText = attribute.value
+                                    break
+
+                                case 'html':
+                                    htmlElement.innerHTML = attribute.value
+                                    break
+
+                                case 'cssClass':
+                                    htmlElement.className = attribute.value
+                                    break
+
+                                default:
+                                    htmlElement.setAttribute(attribute.name, attribute.value)
+                            }
+                        })
+                    }
```

Comment:
obviously users won't try to do this, and it's possible we're making this impossible on the server side, but should we sanitize the contents of these values that we're setting onto the DOM?

Reply:
good idea about sanitizing payload. We can do that on the server-side with nh like we do for survey HTML